storage: when stores are unavailable, don't send the replica to purgatory

BramGruneir opened the pull request:
When a store becomes unavailable due to a declined reservation, we should not
be sending any replicas to purgatory. Purgatory should only be used for when
there are no matching and alive stores.

Fixes #7329

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7332)

<!-- Reviewable:end -->

BramGruneir committed:
storage: add more logging to the replicate queue

BramGruneir commented:
I'm going to spam circle tonight to see if I can get more failures.

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[acceptance/replication_test.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKjAHtu4zJGnZcBo9PC:-KKjAHtu4zJGnZcBo9PD:-1488233377) ([raw file](https://github.com/cockroachdb/cockroach/blob/85abcef1b8fdbeaa6ab0533a0f70c52578db4d6f/acceptance/replication_test.go#L36)):*
> ```Go
> // TODO(bram): Hoping that a longer timeout will fix #7329. Remove this when we
> // find the root cause.
> const waitForReplicationDuration = 30 * time.Second
> ```

Not terribly thrilled about this. @WillHaack was investigating initial replication delays recently. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[acceptance/replication_test.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKjAHtu4zJGnZcBo9PC:-KKjBsJ80bMp00Gvq61q:-1073266412) ([raw file](https://github.com/cockroachdb/cockroach/blob/85abcef1b8fdbeaa6ab0533a0f70c52578db4d6f/acceptance/replication_test.go#L36)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Not terribly thrilled about this. @WillHaack was investigating initial replication delays recently. 
</details>
Also, if the logs weren't helpful - why not add more logging? Are we running the containers at V(1) already? If not, that's a worthwhile change, too. There are some `ChangeReplicas` pushes. If we log enough to, for example, find out that one of the `ChangeReplicas` simply doesn't finish, that would be valuable piece of information.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[acceptance/replication_test.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKjAHtu4zJGnZcBo9PC:-KKjU4S0J7bMxBv1NzQM:1576078062) ([raw file](https://github.com/cockroachdb/cockroach/blob/85abcef1b8fdbeaa6ab0533a0f70c52578db4d6f/acceptance/replication_test.go#L36)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Also, if the logs weren't helpful - why not add more logging? Are we running the containers at V(1) already? If not, that's a worthwhile change, too. There are some `ChangeReplicas` pushes. If we log enough to, for example, find out that one of the `ChangeReplicas` simply doesn't finish, that would be valuable piece of information.
</details>
I'm not happy about it either.

The logs of the containers are now at v1, I made that change the last week.  There was no change replica delays or failures from what I could tell form the logs, but just that range 1 was only replicated once.  All other ranges were fully replicated.  

This doesn't happen all the time, but seems to happen quite often.  And I can't replicate it locally, so I wanted to see if giving it a bit more time to complete would at least cut down on the number of failures.

But I take your point.  I've added some logging to the replicate queue at v1 to gain more insight into the goings on there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
I've been spamming circle.  There should are a number of examples of this failure still happening.  https://circleci.com/gh/BramGruneir/cockroach/tree/7329

I'll be investigating them tomorrow.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-:-KKlTqGuwkbhFqcFDhAK:-1492416922)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
From run #631:

E160621 05:56:11.919336 storage/queue.go:449  [replicate] range=3
[/Table/12-/Table/13): error: storage/replica_command.go:2611: change
replicas of 3 failed: reservation declined:header:<node_id:3
store_id:3 > from_node_id:1 from_store_id:1 range_id:3
range_size:67108864


Not sure if it's relevant. Just happened to be one that I looked at.

On Tue, Jun 21, 2016 at 1:44 AM Bram Gruneir <notifications@github.com>
wrote:

> I've been spamming circle. There should are a number of examples of this
> failure still happening.
> https://circleci.com/gh/BramGruneir/cockroach/tree/7329
>
> I'll be investigating them tomorrow.
> ------------------------------
>
> Review status: 0 of 2 files reviewed at latest revision, 1 unresolved
> discussion, some commit checks pending.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-:-KKlTqGuwkbhFqcFDhAK:-1492416922>*
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7332#issuecomment-227345975>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE135DEQqQOlt_D_80KI3iLjh1zRDp_Rks5qN3ojgaJpZM4I6Ad7>
> .
>
-- 

-- Tobias

petermattis commented:
@BramGruneir There might be other problems with initial replication, but see my latest comment in #7022.

BramGruneir commented:
@tschottdorf That's range 3, not range 1.  So it doesn't cause the slow down in the test.

I'm going to dig through a bunch of the logs today.  Will report back with more.

@petermattis I'm going to see if that will help, I'll rebase off your change and see if we get any repeats of the failures..

petermattis commented:
Looking into acceptance test failures I'm seeing for #7355, I turned on some additional logging and saw:

```
I160621 16:10:40.588263 storage/queue.go:323  [replicate] range=1 [/Min-/Table/11): adding: priority=1000.000
I160621 16:10:40.589059 storage/queue.go:438  [replicate] range=1 [/Min-/Table/11): processing
I160621 16:10:40.589401 storage/queue.go:467  [replicate] range=1 [/Min-/Table/11) (purgatory): error: 0 of 3 stores with an attribute matching []; likely not enough nodes in cluster
```

Note that this occurred after we had replicated ranges `2-5`. Slightly earlier in the log I see a reservation for the replication of range 1 was declined. And then range 1 gets placed in purgatory:

```
I160621 16:10:40.589401 storage/queue.go:467  [replicate] range=1 [/Min-/Table/11) (purgatory): error: 0 of 3 stores with an attribute matching []; likely not enough nodes in cluster
```

When a replica is placed in purgatory it isn't removed until something kicks the `purgatoryChan()`, but there is no outside event to do so in this instance.

tamird commented:
Sounds like we're causing reservation-declined replicas to go into purgatory, which should probably never happen (reservations have their own throttling).

petermattis commented:
Yes, when a reservation is declined we set `storeDetail.unavailableUntil` to a time in the future. We then refuse to match that store so that `StorePool.getStoreList` returns only a subset of the stores in the cluster. This causes the replica that we're trying to be rebalanced to be placed in purgatory from which it can't leave until we receive another gossiped store descriptor. By default, we only gossip store descriptors once per minute.

petermattis commented:
Hacky fix is to set `defaultDeclinedReservationsTimeout` to `0`.

BramGruneir commented:
I'm working on a real fix.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-:-KKnv0B_COA20AxHQeJb:-2144834592)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I'm curious to see what that real fix is.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[acceptance/replication_test.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKjAHtu4zJGnZcBo9PC:-KKnzaqREMjMp6UrmrkR:606580576) ([raw file](https://github.com/cockroachdb/cockroach/blob/85abcef1b8fdbeaa6ab0533a0f70c52578db4d6f/acceptance/replication_test.go#L36)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I'm not happy about it either.
> 
> The logs of the containers are now at v1, I made that change the last week.  There was no change replica delays or failures from what I could tell form the logs, but just that range 1 was only replicated once.  All other ranges were fully replicated.  
> 
> This doesn't happen all the time, but seems to happen quite often.  And I can't replicate it locally, so I wanted to see if giving it a bit more time to complete would at least cut down on the number of failures.
> 
> But I take your point.  I've added some logging to the replicate queue at v1 to gain more insight into the goings on there.
> 
> 
> 
</details>
Btw, given what we know now, this change wouldn't have helped.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-:-KKnzsX6sDHC6qrQZkCX:334331316)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[acceptance/replication_test.go, line 36 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKjAHtu4zJGnZcBo9PC:-KKo1rUdRnfz9xcW_9zf:195102567) ([raw file](https://github.com/cockroachdb/cockroach/blob/85abcef1b8fdbeaa6ab0533a0f70c52578db4d6f/acceptance/replication_test.go#L36)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Btw, given what we know now, this change wouldn't have helped.
</details>
Yeah, I'm going to revert this part, but leave in the extended logging.  It's going to be useful in the future.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332)*
<!-- Sent from Reviewable.io -->

BramGruneir committed:
storage: when stores are unavailable, don't send the replica to purgatory

When a store becomes unavailable due to a declined reservation, we should not
be sending any replicas to purgatory. Purgatory should only be used for when
there are no matching and alive stores.

Fixes #7329

BramGruneir commented:
I've updated the PR with a new commit that should fix the problem.  I'm testing it both locally and in circle to see if it works.

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-:-KKoeO1DgWqx7fNkTu4L:1931208086)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/allocator.go, line 213 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKofyFgdjCmgCYYgvcU:-KKofyFhLGydsQjx12Ke:588975210) ([raw file](https://github.com/cockroachdb/cockroach/blob/d503fbece54382fdc12737274bcc090be33a7fd9/storage/allocator.go#L213)):*
> ```Go
> 			if log.V(1) {
> 				log.Infof("all %d matching stores are currently unavailable", unavailableStoreCount)
> 			}
> ```

`s/all //g`

This is vlogged by `queue.go` as well. I'm not sure if it is not necessary to repeat that logging here.

---

*[storage/store_pool.go, line 99 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKogQsMiLEE45WVQEWb:-KKogQsNr9-8BInMUZ7g:1085384349) ([raw file](https://github.com/cockroachdb/cockroach/blob/d503fbece54382fdc12737274bcc090be33a7fd9/storage/store_pool.go#L99)):*
> ```Go
> 	storeMatchDead        storeMatch = iota // The store is not yet available or has been timed out.
> 	storeMatchAlive                         // The store is alive, but its attributes didn't match the required ones.
> 	storeMatchUnavailable                   // The store is alive and its attributes matched, but it no available.
> ```

`s/but it no available/but unavailable/g`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-:-KKokuVOxYQHsjSTPu_j:1417088721)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/allocator.go, line 213 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKofyFgdjCmgCYYgvcU:-KKonviXYouOMA1HP069:-1535862433) ([raw file](https://github.com/cockroachdb/cockroach/blob/d503fbece54382fdc12737274bcc090be33a7fd9/storage/allocator.go#L213)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/all //g`
> 
> This is vlogged by `queue.go` as well. I'm not sure if it is not necessary to repeat that logging here.
> 
</details>
done. And good point.

---

*[storage/store_pool.go, line 99 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKogQsMiLEE45WVQEWb:-KKoo_7EeMjOC_D9bE_P:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d503fbece54382fdc12737274bcc090be33a7fd9/storage/store_pool.go#L99)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/but it no available/but unavailable/g`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332)*
<!-- Sent from Reviewable.io -->

tamird commented:
I don't see how this prevents stores from getting sent to purgatory, and I don't see a test that looks at that case. Can you add one, please?

---

Reviewed 1 of 1 files at r2, 1 of 3 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[storage/allocator.go, line 211 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKowTM8JVmbguO3FCD5:-KKowTM8JVmbguO3FCD6:2030531432) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/allocator.go#L211)):*
> ```Go
> 		// the replica to purgatory or even consider relaxing the constraints.
> 		if unavailableStoreCount > 0 {
> 			return nil, fmt.Errorf("%d matching stores are currently unavailable", unavailableStoreCount)
> ```

why fmt.Errorf rather than util.Errorf?

---

*[storage/replicate_queue.go, line 138 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKowCS0e8prmOVqBU-t:-KKowCS1R-T9H8r2qlta:-1449561978) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/replicate_queue.go#L138)):*
> ```Go
> 		}
> 		if log.V(1) {
> 			log.Infof("adding replica for under-replicated range:%d to node:%d, store:%d",
> ```

throughout: why don't you just log the replica descriptor?

---

*[storage/store_pool.go, line 99 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKogQsMiLEE45WVQEWb:-KKowmBgbPbuB_4myD29:1822669745) ([raw file](https://github.com/cockroachdb/cockroach/blob/d503fbece54382fdc12737274bcc090be33a7fd9/storage/store_pool.go#L99)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Done.
</details>
more nit (sorry): "but it is unavailable.", otherwise it sounds like the attributes are unavailable.

---

*[storage/store_pool.go, line 100 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKowv2Gu4htv0foXB0-:-KKowv2HYtsxxDGGi2bm:-966714237) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/store_pool.go#L100)):*
> ```Go
> 	storeMatchAlive                         // The store is alive, but its attributes didn't match the required ones.
> 	storeMatchUnavailable                   // The store is alive and its attributes matched, but unavailable.
> 	storeMatchAvailable                     // The store is alive, available and its attributes matched.
> ```

oxford comma!

---

*[storage/store_pool.go, line 390 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKoxHiR7cBH1TGkpeVT:-KKoxHiR7cBH1TGkpeVU:1618789527) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/store_pool.go#L390)):*
> ```Go
> // GetStoreList returns a storeList that contains all active stores that
> // contain the required attributes and their associated stats. It also returns
> // total number of alive stores and the number of unavailable stores.
> ```

missing article

---

*[storage/store_pool.go, line 417 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKoxsI59ZYzNg5o2zUO:-KKoxsI59ZYzNg5o2zUP:-1712064634) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/store_pool.go#L417)):*
> ```Go
> 		case storeMatchAlive:
> 			aliveStoreCount++
> 		case storeMatchUnavailable:
> ```

"unavailable" is too overloaded for this use - even having read the comments on the enum i had to scratch my head to figure out why this case increments both alive and unavailable counts. perhaps "throttling" or some other term that implies that the node is there and responsive but has politely asked not to be a rebalance target.

---

*[storage/store_pool_test.go, line 222 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKoyLKBJhW9_CaaaToj:-KKoyLKBJhW9_CaaaTok:-1300255773) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/store_pool_test.go#L222)):*
> ```Go
> 	sl, aliveStoreCount, unavailableStoreCount := sp.getStoreList(roachpb.Attributes{Attrs: requiredAttrs}, false)
> 	if aliveStoreCount != expectedAliveStoreCount {
> 		return fmt.Errorf("expected AliveStoreCount %d does not match actual %d",
> ```

throughout: maybe these want to be `util.Errorf`s?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-:-KKoyilngUHRyn05n7ZR:-122303972)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Replicas are sent to purgatory when the allocator returns an error that implements the purgatoryError interface.
Interestingly, there was never a test for checking to see if something was destined for purgatory or not.  So I added one.

---

Review status: 1 of 5 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[storage/allocator.go, line 211 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKowTM8JVmbguO3FCD5:-KKt86x3uvVuF1IgG_lE:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/allocator.go#L211)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why fmt.Errorf rather than util.Errorf?
</details>
Done.

---

*[storage/replicate_queue.go, line 138 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKowCS0e8prmOVqBU-t:-KKt8DIKYGcPNao8_HM7:-1491785556) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/replicate_queue.go#L138)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: why don't you just log the replica descriptor?
</details>
There's no need for the extra info.  We already know that the replica is on this store and node.  So really we just need the replica ID and where it's going to (or being removed from).

---

*[storage/store_pool.go, line 99 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKogQsMiLEE45WVQEWb:-KKt8kvwdJ2S7q6aWGQS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d503fbece54382fdc12737274bcc090be33a7fd9/storage/store_pool.go#L99)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> more nit (sorry): "but it is unavailable.", otherwise it sounds like the attributes are unavailable.
</details>
Done.

---

*[storage/store_pool.go, line 100 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKowv2Gu4htv0foXB0-:-KKt2na74X-zyRSQ8hsA:136570262) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/store_pool.go#L100)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> oxford comma!
</details>
I am not a fan of the oxford comma.  :)  If there are ambiguities, one should use semi-colons.

---

*[storage/store_pool.go, line 390 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKoxHiR7cBH1TGkpeVT:-KKtDyOmd9b_SsILvTr5:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/store_pool.go#L390)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> missing article
</details>
Done.

---

*[storage/store_pool.go, line 417 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKoxsI59ZYzNg5o2zUO:-KKt8qNggbL19OHMG039:-1356895687) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/store_pool.go#L417)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "unavailable" is too overloaded for this use - even having read the comments on the enum i had to scratch my head to figure out why this case increments both alive and unavailable counts. perhaps "throttling" or some other term that implies that the node is there and responsive but has politely asked not to be a rebalance target.
</details>
Replaced unavailable with throttled throughout.  That's a much better term.

---

*[storage/store_pool_test.go, line 222 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKoyLKBJhW9_CaaaToj:-KKtDDrRjFd_8Ku3I8e_:-1068804060) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/store_pool_test.go#L222)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: maybe these want to be `util.Errorf`s?
</details>
Indeed.  I think this would be a great case for a new vet.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-:-KKsiZ6ZVu85t5zP3ynv:-1279545181)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 3 of 3 files at r6, 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[storage/allocator_test.go, line 1021 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtfnFh3h7OSadJAH8o:-KKtfnFh3h7OSadJAH8p:-416474310) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1021)):*
> ```Go
> 
> 	// First test to make sure we would send the replica to purgatory.
> 	_, errPurgatory := a.AllocateTarget(
> ```

throughout: plz just use `err`. you never have more than one in scope.

---

*[storage/allocator_test.go, line 1027 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtfbO2FFsU2LsYj1eh:-KKtfbO2FFsU2LsYj1ei:1210065559) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1027)):*
> ```Go
> 		nil,
> 	)
> 	if errPurgatory == nil {
> ```

you don't need this - the check on the next line will catch it.

---

*[storage/allocator_test.go, line 1031 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtfe7NpjbscfkrluT1:-KKtfe7NpjbscfkrluT2:-1917839761) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1031)):*
> ```Go
> 	}
> 	if _, ok := errPurgatory.(purgatoryError); !ok {
> 		t.Fatalf("expected a purgatory error, got: %s", errPurgatory)
> ```

use %v so that nil prints nicely

---

*[storage/allocator_test.go, line 1035 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtfgc7ETMA07DRbnrC:-KKtfgc81VV-IRACjrMd:-976298480) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1035)):*
> ```Go
> 
> 	// Second, test the normal case in which we can allocate to the store.
> 	gossiputil.NewStoreGossiper(g).GossipStores(singleStore, t)
> ```

so this makes allocation succeed? fun.

---

*[storage/allocator_test.go, line 1045 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtfxx9GLLXKPXveQms:-KKtfxx9GLLXKPXveQmt:1643693989) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1045)):*
> ```Go
> 		t.Fatalf("unable to perform allocation: %s", errNone)
> 	}
> 	if result.Node.NodeID != 1 || result.StoreID != 1 {
> ```

isn't this tested elsewhere? doesn't seem useful in this test.

---

*[storage/allocator_test.go, line 1064 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtg6RzPoUGxJYIxuLW:-KKtg6RzPoUGxJYIxuLX:1553962087) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1064)):*
> ```Go
> 		nil,
> 	)
> 	if errNonPurgatory == nil {
> ```

not needed

---

*[storage/allocator_test.go, line 1068 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtg7TheWXSuZ1Y_Go4:-KKtg7TheWXSuZ1Y_Go5:5172406) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1068)):*
> ```Go
> 	}
> 	if _, ok := errNonPurgatory.(purgatoryError); ok {
> 		t.Fatalf("expected a non purgatory error, got: %s", errNonPurgatory)
> ```

%v

---

*[storage/replicate_queue.go, line 138 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKowCS0e8prmOVqBU-t:-KKte_y4VxACgAFUuw-5:220233282) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/replicate_queue.go#L138)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> There's no need for the extra info.  We already know that the replica is on this store and node.  So really we just need the replica ID and where it's going to (or being removed from).
</details>
Huh? You're logging the node and store, and there is no replica ID. I'm suggesting that you log the replica ID so that if more information is added to it in the future, it is naturally logged here (and doesn't require a manual update).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[storage/allocator_test.go, line 1021 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtfnFh3h7OSadJAH8o:-KKtiJy_sbsRCvPmgSgn:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1021)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: plz just use `err`. you never have more than one in scope.
</details>
Done.

---

*[storage/allocator_test.go, line 1027 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtfbO2FFsU2LsYj1eh:-KKtiKofyGM8AgEp51UC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1027)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you don't need this - the check on the next line will catch it.
</details>
Done.

---

*[storage/allocator_test.go, line 1031 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtfe7NpjbscfkrluT1:-KKtiSuLDpx6Js462MKL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1031)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> use %v so that nil prints nicely
</details>
Done.

---

*[storage/allocator_test.go, line 1035 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtfgc7ETMA07DRbnrC:-KKtiU3tgrC2hMdYUJNu:-1747447545) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1035)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> so this makes allocation succeed? fun.
</details>
We need a store in the store pool that's alive and non-throttled.

---

*[storage/allocator_test.go, line 1045 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtfxx9GLLXKPXveQms:-KKti_I98MP-p3fn0ZT_:954692787) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1045)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> isn't this tested elsewhere? doesn't seem useful in this test.
</details>
This is tested elsewhere, but it doesn't hurt to make sure that basic allocation is working.  I'd at least like to keep result != nil, so I figured why not also check out node and store IDs.

---

*[storage/allocator_test.go, line 1064 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtg6RzPoUGxJYIxuLW:-KKtik4ILdCAcw_nyfhb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1064)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not needed
</details>
Done.

---

*[storage/allocator_test.go, line 1068 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtg7TheWXSuZ1Y_Go4:-KKtikXhRpe7nq50Nv2d:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1068)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> %v
</details>
Done.

---

*[storage/replicate_queue.go, line 138 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKowCS0e8prmOVqBU-t:-KKtilhQ_zP8ZCIyQLth:1478547285) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebb8d165181c3a62b61f2689746fb7d4c59794d3/storage/replicate_queue.go#L138)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Huh? You're logging the node and store, and there is no replica ID. I'm suggesting that you log the replica ID so that if more information is added to it in the future, it is naturally logged here (and doesn't require a manual update).
</details>
Ah, I see what you mean.  I meant RangeID in my previous comment.

For add replica, the replica ID is always 0. 

Anyway, added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r8, 5 of 5 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/allocator_test.go, line 1035 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-KKtfgc7ETMA07DRbnrC:-KKtt-1xCPMUJqnVBnh7:-658618358) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb71129e2594dd1ee94489547a3cf5149f96a380/storage/allocator_test.go#L1035)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> We need a store in the store pool that's alive and non-throttled.
</details>
Sure, that's fine, just saying that it's opaque to the lay reader of this test (me).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7332#-:-KKttZHV_nmUEQIytTTw:1417088721)*
<!-- Sent from Reviewable.io -->
